src/psm/src/ir_solver.cpp

```diff
@@ -654,81 +653,120 @@ void IRSolver::createGmatNodes(const vector<dbSBox*>& power_wires,
       // via cut or enclosure and connections cannot be made through
       // enclosures only.
       const int lb = via_bottom_layer->getRoutingLevel();
-      if (bottom_layer_ != lb) {  // do not set for bottom layers
-        Gmat_->setNode(loc, lb);
-      }
+      Gmat_->setNode(loc, lb);
       const int lt = via_top_layer->getRoutingLevel();
-      if (bottom_layer_ != lt) {  // do not set for bottom layers
-        Gmat_->setNode(loc, lt);
-      }
-      // For a stripe we create nodes at the ends of the stripes and at a fixed
-      // frequency in the lowermost layer.
+      Gmat_->setNode(loc, lt);
+    }
+  }
+}
+
+void IRSolver::createGmatWireNodes(const vector<dbSBox*>& power_wires,
+                               const vector<odb::Rect>& macros)
+{
+  for (auto curWire : power_wires) {
+    // For a stripe we create nodes at the ends of the stripes and at a fixed
+    // frequency in the lowermost layer.
+    if (curWire->isVia())
+     continue;
+    dbTechLayer* wire_layer = curWire->getTechLayer();
+    const int l = wire_layer->getRoutingLevel();
+    dbTechLayerDir::Value layer_dir = wire_layer->getDirection();
+    if (l == bottom_layer_) {
+      layer_dir = dbTechLayerDir::Value::HORIZONTAL;
+    }
+    int x_loc1, x_loc2, y_loc1, y_loc2;
+    if (layer_dir == dbTechLayerDir::Value::HORIZONTAL) {
+      y_loc1 = (curWire->yMin() + curWire->yMax()) / 2;
+      y_loc2 = y_loc1;
+      x_loc1 = curWire->xMin();
+      x_loc2 = curWire->xMax();
     } else {
-      dbTechLayer* wire_layer = curWire->getTechLayer();
-      const int l = wire_layer->getRoutingLevel();
-      dbTechLayerDir::Value layer_dir = wire_layer->getDirection();
-      if (l == bottom_layer_) {
-        layer_dir = dbTechLayerDir::Value::HORIZONTAL;
-      }
-      int x_loc1, x_loc2, y_loc1, y_loc2;
+      x_loc1 = (curWire->xMin() + curWire->xMax()) / 2;
+      x_loc2 = x_loc1;
+      y_loc1 = curWire->yMin();
+      y_loc2 = curWire->yMax();
+    }
+    // For all layers we create the end nodes
+    Gmat_->setNode({x_loc1, y_loc1}, l);
+    Gmat_->setNode({x_loc2, y_loc2}, l);
```

Comment:
do we need the end nodes on layer without current sources?

Reply:
I think it's safer to have it. In case two metal stripes in the same layers are connected by overlaps.  The end nodes handle this. If we remove it, it will bring in an assumption that might later lead to false positives in a case where one stripe actually consists of two overlapping wires.